tests/class-two-factor-core.php

```diff
@@ -734,4 +766,85 @@ public function test_show_password_reset_error() {
 
 		$this->assertStringContainsString( 'check your email for instructions on regaining access', $contents );
 	}
+
+	/**
+	 * Validate that a non-2fa login doesn't set the session two-factor data.
+	 *
+	 * @covers Two_Factor_Core::is_current_user_session_two_factor()
+	 * @covers Two_Factor_Core::login_form_validate_2fa()
+	 */
+	public function test_is_current_user_session_two_factor_without_two_factor() {
+		$user = $this->get_dummy_user();
+
+		// Assert user not logged in is false.
+		$this->assertFalse( Two_Factor_Core::is_current_user_session_two_factor() );
+
+		// Set the cookie without going through two-factor, and fill in $_COOKIE.
+		wp_set_auth_cookie( $user->ID );
+
+		$this->assertNotEmpty( self::$last_auth_cookie );
+		$this->assertNotEmpty( self::$last_logged_in_cookie );
+
+		// Simulate the cookies being sent, so that core session functions work.
+		$_COOKIE[ AUTH_COOKIE ]      = self::$last_auth_cookie['auth_cookie'];
+		$_COOKIE[ LOGGED_IN_COOKIE ] = self::$last_logged_in_cookie['logged_in_cookie'];
+
+		// Validate that the session is not flagged as 2FA
+		$this->assertFalse( Two_Factor_Core::is_current_user_session_two_factor() );
+
+		$manager = WP_Session_Tokens::get_instance( $user->ID );
+		$token   = wp_get_session_token();
+		$session = $manager->get( $token );
+
+		// Validate the Session data is not set.
+		$this->assertArrayNotHasKey( 'two-factor-provider', $session );
+	}
+
+	/**
+	 * Validate that a simulated 2fa login sets the session two-factor data.
+	 *
+	 * @covers Two_Factor_Core::is_current_user_session_two_factor()
+	 * @covers Two_Factor_Core::login_form_validate_2fa()
+	 */
+	public function test_is_current_user_session_two_factor_with_two_factor() {
```

Comment:
This only covers about 40% of the function, but a lot of it is "controller" code that we wouldn't test w/ PHPUnit anyway. The suggestion above to break it into a controller/model would make it easier to see what "model" code isn't being tested. That may be more trouble than it's worth, though.

Reply:
This plugin doesn't follow any such design, so kind of have to make do with what we have, and this is that.

Reply:
I didn't mean to imply some kind of formal thing, just basically going farther with what you've already done. But I agree it's not worth trying to do a big refactor like that in this PR 👍🏻